src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -468,11 +462,21 @@ private int WriteCommonFields(Span<byte> buffer, long actualLength, TarEntryType
                 checksum += FormatOctal(_gid, buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
             }
 
-            _size = actualLength;
-
             if (_size > 0)
             {
-                checksum += FormatOctal(_size, buffer.Slice(FieldLocations.Size, FieldLengths.Size));
+                if (_size <= TarHelpers.MaxSizeLength)
+                {
+                    checksum += FormatOctal(_size, buffer.Slice(FieldLocations.Size, FieldLengths.Size));
+                }
+                else if (_format is not TarEntryFormat.Pax)
+                {
+                    throw new ArgumentException(SR.Format(SR.TarSizeFieldTooLargeForEntryFormat, _format));
+                }
+                else
+                {
+                    Debug.Assert(_typeFlag is not TarEntryType.ExtendedAttributes);
```

Comment:
Makes sense to avoid writing 8GB of extended attributes. I would never expect that to happen.

I would also add GlobalExtendedAttributes here:

```suggestion
                    Debug.Assert(_typeFlag is not TarEntryType.ExtendedAttributes and not TarEntryType.GlobalExtendedAttributes);
```

Reply:
(Because Global Extended Attributes entries have a pax format too)